benches/bigint_arithmetic.rs

```diff
@@ -2,26 +2,22 @@ use criterion::{black_box, Criterion};
 use criterion::criterion_group;
 use criterion::criterion_main;
 
-use plonky::{cmp_6_6, mul_6_6};
+use plonky::{cmp, mul};
```

Comment:
I'm seeing a couple errors here -- it looks like `cmp` is not visible in this context, and `mul` is missing? (A normal `cargo build` wouldn't detect these, but we can use `cargo bench --no-run` to build the benchmark code.)

Re visibility, I think the compiler sees benchmarks as a separate crate, so AFAIK the only easy solution is to make `cmp` `pub`, though it's not ideal. Hopefully the compiler will change that at some point.

Reply:
Good catch! Quite right, I was just `cargo test`ing so I missed these errors. (I removed `mul` because the only time it was used was in the benchmarking code.)

And you're right about the visibility: In the previous version, `cmp_4_4` was declared `pub (crate)` but `cmp_6_6` was declared `pub`; in this refactor I used `pub (crate)` which broke the access to `cmp` in the benchmark.

Given that we would prefer not to have `cmp` public, does it make sense to put the benchmark in the crate somehow? The benchmark currently only has `cmp` tests, so it's pretty sparse; presumably many functions that we will want to benchmark in the future shouldn't be `pub` either.

Reply:
I agree having benchmarks inside the crate would be cleaner, but it seems it isn't currently possible with Criterion.rs, as per https://github.com/bheisler/criterion.rs/issues/158 --

> Cargo does not expose any way for Criterion.rs to define benchmarks in the main crate (ie. in the src/ directory).

It seems like our options are
- Switch from Criterion.rs to Rust's native benchmark framework, though it's nightly-only
- Move functions like `cmp` to a separate crate, so they can be `pub` without appearing in Plonky's docs
- Just make the functions `pub` for now until something changes, like native benchmarking becoming stable
- Make these functions public only if a feature is enabled, as suggested [here](https://www.reddit.com/r/rust/comments/9rnuhi/what_should_i_use_today_for_benchmarking_in_rust/e8ii3pk/?utm_source=reddit&utm_medium=web2x&context=3)

I don't really like the last option, since new contributors would have to consult the readme to know how to run benchmarks, vs just running `cargo bench`. The others all seem reasonable to me -- what do you think? (Also @wborgeaud -- let us know if you have a preference.)

Reply:
I think the easiest way forward is to take option 3: make `cmp` (and perhaps similar functions) public for now; we can keep an eye on improvements to Criterion and Rust's native benchmark and fix the visibility when they catch up.

In fact, I suspect that most (all?) of the `bigint_arithmetic` functions are benchmarked just as well indirectly via the public interfaces of the various concrete field types. For example, there's little point in benchmarking `bigint_arithmetic::add_no_overflow` since it's covered by the various `Add` trait implementations for Tweedle{dee,dum}, etc. So one option is to remove the `bigint_arithmetic` benchmarks altogether. Thoughts?

Reply:
Hm in theory, if we were optimizing a private helper method, it seems like a bench that only calls that method might give us the best signal-to-noise ratio. Though in practice, for tiny methods like `cmp`, maybe that bench data could be misleading? Since we wouldn't be seeing how changes to `cmp` affect the pipelines and register allocations of callers.

I agree option 3 sounds good so that we can quickly unblock your change, and then let's decide on the more general convention in our next call.

Reply:
Ok, committed option 3 solution. Yep, let's discuss next call.